File: Leetcode75/Queue/649_Dota2_Senate.cpp
#include <iostream>
#include <queue>
using namespace std;

/*
    Time: O(N), Space: O(N)
*/

string predictPartyVictory(string senate) {
    queue<int> radiant;
    queue<int> dire;
    for (int i = 0; i < senate.size(); i++) {
        if (senate[i] == 'R') {
            radiant.push(i);
        } else {
            dire.push(i);
        }
    }

    int index = senate.size();
    while (!radiant.empty() && !dire.empty()) {
        index++;
        if (radiant.front() < dire.front()) {
            radiant.push(index);
        } else {
            dire.push(index);
        }
        radiant.pop(), dire.pop();
    }

    if (radiant.empty()) {
        return "Dire";
    } else {
        return "Radiant";
    }
}

File: Javascript/Queue/649_Dota2_Senate.js
/**
 * @param {string} senate
 * @return {string}
 */
var predictPartyVictory = function (senate) {
    const d = [];
    const r = [];
    for (let i = 0; i < senate.length; i++) {
        if (senate[i] === 'D') d.push(i);
        else r.push(i);
    }

    const n = senate.length;
    while (d.length > 0 && r.length > 0) {
        const senatorD = d.shift();
        const senatorR = r.shift();
        if (senatorD < senatorR) {
            // dire first
            d.push(n + senatorD);
        } else {
            r.push(n + senatorR);
        }
    }

    return (d.length === 0) ? "Radiant" : "Dire";
};